Yggdrasil (Mercenary Overlord)

SirChuckOfTheChuckles opened the pull request:
Generator PR: https://github.com/Ziktofel/Archipelago/pull/334

Adds the Yggdrasil Overlord Mercenary, summonable from the Predator Nest in the Brood War mercenary tab.

### Basic Stats:
Cost: 250 Minerals
Grants 30 Supply
Cooldown Time: 240 seconds
Health: 500
Armor: 2
Attributes: Biological, Armored, Massive

### Abilities
Has same upgrades as regular overlords, but cannot morph into an overseer. When generating creep, generates it at a faster speed than regular overlords.

https://github.com/user-attachments/assets/3adc5470-88f5-43c0-94b2-5422039e404b


**NEW ABILITY: ASSIMILATE** (Default Hotkey T): Targets friendly zerg structures (specifically ones morphable from drones), destroying them but giving the Yggdrasil a buff corresponding to the structure destroyed. The Yggdrasil can then recast the ability to build the structure, at no cost and at a reduced build time. The structure does not provide tech bonuses while it's life essence is within the Yggdrasil.


https://github.com/user-attachments/assets/514810b5-73df-4c13-8dde-faeca46657be


https://github.com/user-attachments/assets/1ac14db7-3b03-4997-9feb-8596af87da6a



https://github.com/user-attachments/assets/d306c012-bc28-48cd-94bc-c098697b74d2




Current list of structures supported:
- Hatchery  
- Lair  
- Hive  
- Extractor  
- Automated Extractor  
- Hydralisk Den  
- Greater Hydralisk Den  
- Spire  
- Greater Spire  
- Baneling Nest  
- Evolution Chamber  
- Spawning Pool   
- Roach Warren  
- Spine Crawler  
- Spore Crawler  
- Infestation Pit  
- Nydus Network  
- Ultralisk Cavern  
- Predator Nest

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/ActorData.xml:
```diff
@@ -41782,4 +41792,136 @@
+        <HostSiteOps Ops="SOpAttachOrigin"/>
+        <Model value="ConsumptionImpact"/>
+    </CActorModel>
+    <CActorUnit id="AP_Yggdrasil" parent="GenericUnitMorph" unitName="AP_Overlord">
+        <Aliases value="_UnitLarge"/>
+        <Macros value="PhysicsDeathsSwarmAir"/>
+        <Macros value="PhysicsDeathsVoidAir"/>
+        <On index="0" Terms="UnitBirth.AP_Yggdrasil"/>
+        <On index="1" Terms="UnitBirth.AP_Yggdrasil"/>
+        <On index="2" Terms="UnitRevive.AP_Yggdrasil"/>
+        <On index="3" Terms="UnitRevive.AP_Yggdrasil"/>
+        <On index="4" Terms="UnitConstruction.AP_Yggdrasil.Start"/>
+        <On index="5" Terms="UnitConstruction.AP_Yggdrasil.Finish"/>
```
If you set `unitName` to `AP_Yggdrasil`, you should be able to simplify a lot

SirChuckOfTheChuckles replied:
Cleaned up some of the junk, specifically "On index = "x" " lines.

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/enUS.SC2Data/LocalizedData/GameHotkeys.txt:
```diff
@@ -587,3 +587,25 @@ Button/Hotkey/AP_BansheeMengskNightAssaultOff=T
 Button/Hotkey/AP_KerriganVoidCoopCrushingGripWave=Z
+Button/Hotkey/AP_ExterminationProtocol=V
+Button/Hotkey/AP_SummonYggdrasil=O
+Button/Hotkey/AP_YggdrasilAssimilate=T
+Button/Hotkey/AP_YggdrasilEvolutionChamber=T    
+Button/Hotkey/AP_YggdrasilGreaterSpire=T
+Button/Hotkey/AP_YggdrasilHatchery=T
+Button/Hotkey/AP_YggdrasilHive=T
+Button/Hotkey/AP_YggdrasilHydraliskDen=T
+Button/Hotkey/AP_YggdrasilInfestationPit=T
+Button/Hotkey/AP_YggdrasilLair=T
+Button/Hotkey/AP_YggdrasilLurkerDen=T
+Button/Hotkey/AP_YggdrasilNydusNetwork=T
+Button/Hotkey/AP_YggdrasilPredatorNest=T
+Button/Hotkey/AP_YggdrasilRoachWarren=T
+Button/Hotkey/AP_YggdrasilSpawningPool=T
+Button/Hotkey/AP_YggdrasilSpineCrawler=T
+Button/Hotkey/AP_YggdrasilSpire=T
+Button/Hotkey/AP_YggdrasilSporeCrawler=T
+Button/Hotkey/AP_YggdrasilUltraliskCavern=T
+Button/Hotkey/AP_YggdrasilBanelingNest=T
+Button/Hotkey/AP_YggdrasilExtractor=T
+Button/Hotkey/AP_YggdrasilAutomatedExtractor=T
```
Link the hotkeys together as they are in one slot and you want those rebindable in the hotkey editor (the build buttons should have hotkey entry of `Button/Hotkey/AP_YggdrasilAssimilate`)

SirChuckOfTheChuckles replied:
Should be fixed within the buttons section.

SirChuckOfTheChuckles commented:
Made following bug fixes:
Fixed an issue where Yggdrasil could be unusuable after right-clicking a building under construction.
Fixed an issue where Fast Delivery upgrade (no initial time for mercenaries) did not work with Yggdrasil.
